### PR TITLE
[keymgr/dv] Support sideload clear

### PR DESCRIPTION
**Pull request by weicaiyang**
In sequences, randomize `sideload_clear` after gen-hw-out
In checkers, check each sideload interface key not matching if it's
cleared

Signed-off-by: Weicai Yang <weicai@google.com>

### CODE REVIEW

**Review by tjaychen — APPROVED** [review]
lgtm!

**Review by cindychip — APPROVED** [review]

File: hw/ip/keymgr/dv/env/keymgr_if.sv [hunk 1]
```diff
@@ -8,6 +8,17 @@ interface keymgr_if(input clk, input rst_n);
   import uvm_pkg::*;
   import keymgr_env_pkg::*;
 
+  // Initially, status should be SideLoadNotAvail, after sideload key is generated, it becomes
+  // SideLoadClear.
+  // status can't be directly changed from SideLoadClear to SideLoadAvail.
+  // When status is SideLoadClear due to SIDELOAD_CLEAR programmed, need to write CSR to 0 to reset
+  // it so that status is changed to SideLoadNotAvail, then we may set it to SideLoadAvail again
```
Comment (sriyerg) [inline hw/ip/keymgr/dv/env/keymgr_if.sv:15]: nit: Probably better to explain what it is first, rather than jumping straight to the legality of the enum transitions. 
```
  // Represents the keymgr sideload state for each sideload interface.
  //
  // The initial status is SideLoadNotAvail. After the sideload key is generated ...
```

File: hw/ip/keymgr/dv/env/keymgr_if.sv [hunk 2]
```diff
@@ -34,13 +45,16 @@ interface keymgr_if(input clk, input rst_n);
...
   bit is_kmac_key_good;
-  bit is_aes_key_good;
-  bit is_otbn_key_good;
+
+  // sideload status
+  keymgr_sideload_status_e aes_sideload_status;
+  keymgr_sideload_status_e otbn_sideload_status;
 
   // when kmac sideload key is generated, kmac may be used to do other OP, but once the OP is done,
   // it should automatically switch back to sideload key
```
Comment (sriyerg) [inline hw/ip/keymgr/dv/env/keymgr_if.sv:56]: nit: How about this:
```
  // KMAC may be busy with other operations when the keymgr initiates a sideload key generation with it. 
  // But once it is done, it is expected to switch automatically to process the keymgr sideload request. 
```

File: hw/ip/keymgr/dv/env/keymgr_if.sv [hunk 3]
```diff
@@ -85,9 +99,9 @@ interface keymgr_if(input clk, input rst_n);
     aes_key_exp  = '0;
     otbn_key_exp = '0;
     is_kmac_key_good = 0;
-    is_aes_key_good  = 0;
-    is_otbn_key_good = 0;
-    is_kmac_sideload_avail = 0;
+    kmac_sideload_status = SideLoadNotAvail;
+    aes_sideload_status = SideLoadNotAvail;
+    otbn_sideload_status = SideLoadNotAvail;
```
Comment (sriyerg) [inline hw/ip/keymgr/dv/env/keymgr_if.sv:104]: Should we check that these are back to `SideLoadNotAvail` state in `check_phase`?

File: hw/ip/keymgr/dv/env/keymgr_if.sv [hunk 4]
```diff
@@ -284,12 +344,19 @@ interface keymgr_if(input clk, input rst_n);
...
-        @(posedge clk);
+        @(kmac_key or is_kmac_key_good);
+        #1ps; // avoid race condition
         if (!is_kmac_key_good) check_invalid_key(kmac_key, "KMAC");
       end
       forever begin
-        @(posedge clk);
-        if (!is_aes_key_good) check_invalid_key(aes_key, "AES");
+        @(aes_key or aes_sideload_status);
+        #1ps; // avoid race condition
```
Comment (sriyerg) [inline hw/ip/keymgr/dv/env/keymgr_if.sv:353]: Not sure I understand why is this pound delay needed? You would definitely not need it if you used a clocking block and made NBAs in the interface when driving the DUT. 

Could you also try `forever @(foo, bar) begin` style?

File: hw/ip/keymgr/dv/env/keymgr_scoreboard.sv [hunk 5]
```diff
@@ -547,8 +547,13 @@ class keymgr_scoreboard extends cip_base_scoreboard #(
...
       end
       "reseed_interval_shadowed": begin
-        if (addr_phase_write)
-          cfg.keymgr_vif.edn_interval = `gmv(ral.reseed_interval_shadowed.val);
+        if (addr_phase_write) cfg.keymgr_vif.edn_interval = `gmv(ral.reseed_interval_shadowed.val);
+      end
+      "sideload_clear": begin
+        if (addr_phase_write) begin
+          cfg.clk_rst_vif.wait_clks(1);
+          cfg.keymgr_vif.clear_sideload_key(`gmv(ral.sideload_clear.val));
```
Comment (sriyerg) [inline hw/ip/keymgr/dv/env/keymgr_scoreboard.sv:555]: Should you be forking this off to a separate thread? Not sure if its a good idea to have `process_tl_access` consume time.

File: hw/ip/keymgr/dv/env/keymgr_if.sv (hunk 1 above)
Comment (weicaiyang) [reply hw/ip/keymgr/dv/env/keymgr_if.sv]: Updated, thanks.

File: hw/ip/keymgr/dv/env/keymgr_if.sv (hunk 2 above)
Comment (weicaiyang) [reply hw/ip/keymgr/dv/env/keymgr_if.sv]: it doesn't mean sideload request is interrupted. Let me rephrase to make it clear.

File: hw/ip/keymgr/dv/env/keymgr_if.sv (hunk 3 above)
Comment (weicaiyang) [reply hw/ip/keymgr/dv/env/keymgr_if.sv:109]: We can't do that as it can be any of those 3 values.

Comment (sriyerg) [reply hw/ip/keymgr/dv/env/keymgr_if.sv:110]: I am used to bringing IP back to a known state at the end (which was the motivation of `dut_shutdown()` added to `dv_base_vseq`) so that we can check if the expected state we maintain in our testbench matches the DUT. It is probably ok I suppose.

File: hw/ip/keymgr/dv/env/seq_lib/keymgr_sideload_vseq.sv [hunk 6]
```diff
@@ -7,8 +7,32 @@ class keymgr_sideload_vseq extends keymgr_smoke_vseq;
...
+
   // also test HW sideload
   constraint gen_operation_c {
     gen_operation inside {keymgr_pkg::OpGenId, keymgr_pkg::OpGenSwOut, keymgr_pkg::OpGenHwOut};
   }
+
+  constraint clear_dest_c {
+    clear_dest dist {[0:3] :/ 4,
+                     // reserved value, clear all sideload
+                     [4:$] :/ 1};
```
Comment (sriyerg) [inline hw/ip/keymgr/dv/env/seq_lib/keymgr_sideload_vseq.sv:21]: Not sure I understand what this is 3 bits, if there are only 4 possible values (0-3). Can it be 2 bits instead?
https://docs.opentitan.org/hw/ip/keymgr/doc/index.html#Reg_sideload_clear
@tjaychen

Comment (weicaiyang) [reply hw/ip/keymgr/dv/env/seq_lib/keymgr_sideload_vseq.sv:21]: I think we do need 3 bits, so that we can these options.

> 0: not clear, which can be used to change from clear to not clear
> 1, 2, 3: clear one of 3 sideload interfaces
> 4 or others: clear all 3 sideload interfaces

File: hw/ip/keymgr/dv/env/keymgr_if.sv (hunk 3 above)
Comment (weicaiyang) [reply hw/ip/keymgr/dv/env/keymgr_if.sv:110]: sequence will make sure the sideload operation is completed, otherwise, it finishes at an unknown state and we have unfinished item in the queue . For the generated sideload data, it's stable and I think it's fine whether we clear it or keep it available for other block. (it's similar to testing pwrmgr, we turn on the power for blocks, but probably don't need to turn them off before exiting the test)

Comment (sriyerg) [reply hw/ip/keymgr/dv/env/keymgr_if.sv:110]: Thanks for the explanation.

File: hw/ip/keymgr/dv/env/seq_lib/keymgr_sideload_vseq.sv (hunk 6 above)
Comment (sriyerg) [reply hw/ip/keymgr/dv/env/seq_lib/keymgr_sideload_vseq.sv:21]: I see, ok.

File: hw/ip/keymgr/dv/env/keymgr_if.sv (hunk 4 above)
Comment (sriyerg) [reply hw/ip/keymgr/dv/env/keymgr_if.sv:359]: I think there is some semantic difference - I don't recall. Will check the LRM. This is fine for now.

**Review by sriyerg — APPROVED** [review]
LGTM, thanks!
